flowapi/flowapi/permissions.py

```diff
@@ -1,245 +1,64 @@
+    if isinstance(in_iter, dict):
+        in_iter = in_iter.values()
+    # Think there's a slightly better way of doing type introspection here
+    return all(type(item) not in [dict, list] for item in in_iter)
 
 
 @functools.singledispatch
-def valid_tree_walks(
-    tree,
-    *,
-    paths: Optional[Iterable[str]] = None,
-    depth: int = 1,
-) -> Union[tuple, list]:
-    """
-
-    Parameters
-    ----------
-    tree : list of str or dict
-        Tree to walk through
-    paths : iterable of str or None
-        Nodes already traversed
-    depth : int, default 1
-        Current depth in the tree
-
-
-    Yields
-    ------
-    list of str or tuple of list of str
-        Path through the tree
-
-    """
-    yield [*paths, tree]
-
-
-@valid_tree_walks.register
-def _(
-    tree: list,
-    *,
-    paths: Optional[Iterable[str]] = None,
-    depth: int = 1,
-) -> Union[tuple, list]:
-    if paths is None:
-        paths = tuple()
-    if len(tree) == 0 and len(paths) > 0:
-        yield list(paths)
-    else:
-        for v in tree:
-            yield [*paths, v]
-
-
-@valid_tree_walks.register
-def _(
-    tree: dict,
-    *,
-    paths: Optional[Iterable[str]] = None,
-    depth: int = 1,
-) -> Union[tuple, list]:
-    even_depth = depth % 2 == 0
-    if paths is None:
-        paths = tuple()
-    if len(tree) == 0 and len(paths) > 0:
-        yield list(paths)
-    elif not even_depth or len(tree) == 1:
-        for k, v in sorted(tree.items(), key=dict_and_list_key_function):
-            yield from valid_tree_walks(v, paths=(*paths, k), depth=depth + 1)
-    else:
-        yield from product(
-            *(
-                valid_tree_walks(v, paths=(*paths, k), depth=depth + 1)
-                for k, v in sorted(
-                    tree.items(),
-                    key=dict_and_list_key_function,
-                )
-            )
-        )
-
-
-@functools.singledispatch
-def tree_walk_to_scope_list(tree_walk) -> str:
-    """
-
-    Parameters
-    ----------
-    tree_walk : str, list or str or tuple of list of str
-        A path through the tree
-
-    Yields
-    ------
-    str
-        The tree walk as a . delimited string
-
-    """
-    yield tree_walk
+def _flatten_on_key_inner(root, key_of_interest):
+    raise TypeError
 
 
-@tree_walk_to_scope_list.register
-def _(tree_walk: list) -> str:
-    yield ".".join(tree_walk)
-
-
-@tree_walk_to_scope_list.register
-def _(tree_walk: tuple) -> str:
-    for walk in tree_walk:
-        yield from tree_walk_to_scope_list(walk)
-
+@_flatten_on_key_inner.register(dict)
+def _(root, key_of_interest):
+    for node, value in root.items():
+        if is_flat(value):
+            pass
+        else:
+            yield from _flatten_on_key_inner(value, key_of_interest)
+            if node == key_of_interest:
+                # We cannot change the size of a dict mid-iterate, so instead we mark it for
+                # deletion post-iterate
+                root[node] = {}
```

Comment:
Oo. I don't love the mutation. Must be a better way?

Reply:
Yeah, I was racking my brains about this for a while - I'm going to circle back around to this once I've got the rest of the API in place, but there's gotta be a way of marking as part of the coroutine instead of the dict itself